cmd/apiserver/main.go

```diff
@@ -47,158 +41,15 @@ func gatherOptions() options {
-	}
-}
-
-func (o *options) simulateRun(ctx context.Context, fakeClient v1.VegaV1Interface, calcName string, logger *logrus.Entry) {
-	logger.Info("Starting simulation")
-
-	ticker := time.NewTicker(time.Duration(o.dryTickerMinutes) * time.Minute)
-	defer ticker.Stop()
-	done := make(chan bool)
-
-	go func() {
-		opts := metav1.SingleObject(metav1.ObjectMeta{Name: calcName})
-		watcher, _ := fakeClient.Calculations().Watch(ctx, opts)
-		defer watcher.Stop()
-
-		// Watch calculation until Completed or Failed status
-		for {
-			select {
-			case event, _ := <-watcher.ResultChan():
-				obj := event.Object.(*calculationsv1.Calculation)
-				if obj.Phase == calculationsv1.CompletedPhase || obj.Phase == calculationsv1.FailedPhase {
-					done <- true
-				}
-			}
-		}
-	}()
-
-	getPhaseWithFailureChance := func(chance int) calculationsv1.CalculationPhase {
-		r := rand.Intn(100)
-		if r < chance {
-			return calculationsv1.FailedPhase
-		}
-		return calculationsv1.CompletedPhase
-	}
-
-	isFinished := func(spec calculationsv1.CalculationSpec) bool {
-		for _, step := range spec.Steps {
-			if step.Status == calculationsv1.ProcessingPhase || step.Status == calculationsv1.CreatedPhase {
-				return false
-			}
-		}
-		return true
-	}
-
-	for {
-		select {
-		case <-done:
-			logger.Warn("Simulation finished")
-			return
-		case <-ticker.C:
-			newCalc, err := fakeClient.Calculations().Get(ctx, calcName, metav1.GetOptions{})
-
-			if err != nil {
-				logger.WithError(err).Error("couldn't get calculation")
-				goto End
-			}
-
-			switch newCalc.Phase {
-			case calculationsv1.CreatedPhase:
-				newCalc.Phase = calculationsv1.ProcessingPhase
-				newCalc.Spec.Steps[0].Status = calculationsv1.ProcessingPhase
-				break
-
-			case calculationsv1.ProcessingPhase:
-				if isFinished(newCalc.Spec) {
-					newCalc.Phase = calculationsv1.CompletedPhase
-				}
-
-				for index, step := range newCalc.Spec.Steps {
-					switch step.Status {
-
-					case calculationsv1.CompletedPhase:
-						continue
-
-					case calculationsv1.FailedPhase:
-						newCalc.Phase = calculationsv1.FailedPhase
-						goto End
-
-					case calculationsv1.ProcessingPhase:
-						newCalc.Spec.Steps[index].Status = getPhaseWithFailureChance(o.dryRunFailureRate)
-						goto End
-
-					case calculationsv1.CreatedPhase:
-						newCalc.Spec.Steps[index].Status = calculationsv1.ProcessingPhase
-						goto End
-					}
-				}
-			}
-
-		End:
-			logger.Info("Updating calculation")
-			fakeClient.Calculations().Update(ctx, newCalc, metav1.UpdateOptions{})
-		}
-	}
-}
-
 func main() {
-	o := gatherOptions()
-
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
+	o := Bind(flag.NewFlagSet(os.Args[0], flag.ExitOnError))
+	sim := Simulator{
+		ctx:                  context.Background(),
+		dryCalculationsTotal: o.dryCalculationsTotal,
+		dryRunFailureRate:    o.dryRunFailureRate,
+		dryWorkers:           o.dryWorkers,
+		dryTickerMinutes:     o.dryTickerMinutes,
+	}
```

Comment:
This piece of code should be replaced by Bind method from simulator?

Reply:
yes

Reply:
What would `o.simulator.Bind(fs)` do in this case? Don't I want to do that only if `dryRun` is true? I should add simulator struct to options struct? 
I am sorry, but I don't see why is this needed here. I thought this would be sufficient.
Also the Bind method look like this, which I am not sure at all with this: 
`func (s *simulator) Bind(fs *flag.FlagSet) simulator {
	fs.IntVar(&s.dryCalculationsTotal, "dry-total-calculations", 100, "Number of total calculations (dry-run)")
	fs.IntVar(&s.dryRunFailureRate, "dry-failure-rate", 20, "Calculations failure rate in percentage (dry-run)")
	fs.IntVar(&s.dryWorkers, "dry-workers", 10, "Number of workers (dry-run)")
	fs.IntVar(&s.dryTickerMinutes, "dry-ticker", 1, "Minutes per calculation update (dry-run)")
	return *s
}`

Also in main, when dryRun is true, I end up doing this 
           `sim := simulator{
			ctx: ctx,
		}
		sim = sim.Bind(flag.NewFlagSet(os.Args[0], flag.ExitOnError))`
which looks completely weird to me, because the `ctx` has to be determined here, before I call the `dryRun` functions, where ctx is needed.

Reply:
I think you misunderstood the reason for the `Bind` method. First, this shouldn't return a `simulator` at all. The purpose of this is to bind the flags that are related to the simulator with the main flags.

Reply:
The `o.simulator.Bind(fs)` will bind the simulator flags to the main flags. The same way we have it now, but the values would exist in the simulator struct instead of `options`.

Reply:
Well, I tried to do something about it, but I still don't get the reasoning behind it at all. Doesn't make sense to me..

Reply:
All of those options are related only to the simulator, but we also need to control them via flags. In this way, you don't need to pass their values between structs as you did before.